ember-cli module directory emptied of files

When using a `yarn.lock` but without bower as a dependency in `package.json`, the `node_modules/ember-cli/` is emptied of files.

* Failed build with error: https://circleci.com/gh/Vestorly/torii/614
* Successful build: https://circleci.com/gh/Vestorly/torii/615 
* Only change to fix was adding bower to `package.json`: https://github.com/Vestorly/torii/pull/382/commits/43d9737b617a5a555b5d325e23fe7ad23961997f

Currently when `yarn` is detected ember-try attempts to use it for most operations. An exception is that if bower is not a dependency, `npm` is used to install that dependency. I presume there is some negative interaction between the two systems that causes the files to be deleted.

I got this error while running tests with ember try:

> The command ember t exited Error: Cannot find module 'ember-cli'

I am on `5.3.0` npm and do not want to downgrade npm. Could my error possibly be related to this issue?

--- a/all-commands.sh
+++ b/all-commands.sh
@@ -31,46 +31,11 @@ ember try:ember '1.13.0' --skip-cleanup=true
 ember try:config
 ember try:config --config-path='test/fixtures/dummy-ember-try-config.js'
 
-# try:testall (Deprecated)
-ember try:testall
-
-# all styles of options for ember-try's own option
-ember try:testall --skip-cleanup
-ember try:testall --skip-cleanup=true
-ember try:testall --skip-cleanup true
-
 # config-path option
 ember try:testall --config-path='test/fixtures/dummy-ember-try-config.js'
 
 # both ember-try options
 ember try:testall --config-path='test/fixtures/dummy-ember-try-config.js' --skip-cleanup true
-
-# try <scenario>  (Deprecated)
-ember try default
-
-# custom command
-ember try default help
-
-# skip-cleanup option
-ember try default --skip-cleanup
-
-# The following two DO NOT CURRENTLY WORK
-# ember try default --skip-cleanup=true
-# ember try default --skip-cleanup true
-
-# config-path option; DOES NOT CURRENTLY WORK
-# ember try test1 --config-path='test/fixtures/dummy-ember-try-config.js'
-
-# both ember-try options; DOES NOT CURRENTLY WORK
-# ember try test1 --config-path='test/fixtures/dummy-ember-try-config.js' --skip-cleanup true
-
-# custom command with all styles of options
-ember try default help --silent
-ember try default help --silent=true
-ember try default help --silent true
-
-# custom command mixed with ember try's own option
-ember try default help --silent --skip-cleanup
 
 # try:one <scenario>
 ember try:one default
@@ -96,7 +61,6 @@ ember try:one default --- ember help --silent true
 
 # custom command mixed with ember try's own option
 ember try:one default --skip-cleanup --- ember help --silent
-
 
 # try:reset
 ember try:reset

--- a/lib/dependency-manager-adapters/bower.js
+++ b/lib/dependency-manager-adapters/bower.js
@@ -121,17 +121,6 @@ module.exports = CoreObject.extend({
     return adapter._findBowerPath(adapter.cwd).then(function(bowerPath) {
       debug('Run bower install using bower at %s', bowerPath);
       return adapter.run('node', [].concat([bowerPath], commandParts, options), { cwd: adapter.cwd });
-    }).catch(function() {
-      debug('Local bower not found');
-      return adapter._checkForGlobalBower().then(function() {
-        debug('Using global bower');
-        return adapter.run('bower', [].concat(commandParts, options), { cwd: adapter.cwd });
-      });
-    });
-  },
-  _checkForGlobalBower: function() {
-    return this.run('bower', ['-v'], { stdio: 'ignore' }).catch(function() {
-      throw new Error('Bower must be installed either locally or globally to have bower scenarios');
     });
   },
   _bowerJSONForDependencySet: function(bowerJSON, depSet) {

--- a/lib/dependency-manager-adapters/bower.js
+++ b/lib/dependency-manager-adapters/bower.js
@@ -108,17 +108,31 @@ module.exports = CoreObject.extend({
   },
   _install: function() {
     var adapter = this;
-    var options = this.managerOptions || [];
-
     return rimraf(path.join(adapter.cwd, 'bower_components'))
       .then(function() {
-        debug('Remove bower_components');
-        return adapter._findBowerPath(adapter.cwd);
-      })
-      .then(function(bowerPath) {
-        debug('Run bower install using bower at %s', bowerPath);
-        return adapter.run('node', [].concat([bowerPath, 'install', '--config.interactive=false'], options), { cwd: adapter.cwd });
+        debug('Removed bower_components');
+        return adapter._runBowerInstall();
       });
+  },
+  _runBowerInstall: function() {
+    var adapter = this;
+    var options = this.managerOptions || [];
+    var commandParts = ['install', '--config.interactive=false'];
+    return adapter._findBowerPath(adapter.cwd).then(function(bowerPath) {
+      debug('Run bower install using bower at %s', bowerPath);
+      return adapter.run('node', [].concat([bowerPath], commandParts, options), { cwd: adapter.cwd });
+    }).catch(function() {
+      debug('Local bower not found');
+      return adapter._checkForGlobalBower().then(function() {
+        debug('Using global bower');
+        return adapter.run('bower', [].concat(commandParts, options), { cwd: adapter.cwd });
+      });
+    });
+  },
+  _checkForGlobalBower: function() {
+    return this.run('bower', ['-v'], { stdio: 'ignore' }).catch(function() {
+      throw new Error('Bower must be installed either locally or globally to have bower scenarios');
+    });
   },
   _bowerJSONForDependencySet: function(bowerJSON, depSet) {
     if (!bowerJSON.resolutions) {
@@ -172,26 +186,13 @@ module.exports = CoreObject.extend({
   },
   _findBowerPath: function() {
     var adapter = this;
-    return findEmberPath(adapter.cwd)
-      .then(function(emberPath) {
-        /* Find bower's entry point module relative to
-         ember-cli's entry point script */
-        return adapter._resolveModule('bower', { basedir: path.dirname(emberPath) })
-          .catch(function() {
-            debug('Bower not found');
-            return adapter._installBower();
-          }).then(function() {
-            return adapter._resolveModule('bower', { basedir: path.dirname(emberPath) });
-          });
-      })
-      .then(function(bowerPath) {
-        return '"' + path.join(bowerPath, '..', '..', 'bin', 'bower') + '"';
-      });
-  },
-  _installBower: function() {
-    var adapter = this;
-    debug('Installing bower via npm');
-    return adapter.run('npm', [].concat(['install', 'bower@^1.3.12']), { cwd: adapter.cwd });
+    return findEmberPath(adapter.cwd).then(function(emberPath) {
+      /* Find bower's entry point module relative to ember-cli's entry point script */
+      return adapter._resolveModule('bower', { basedir: path.dirname(emberPath) })
+        .then(function(bowerPath) {
+          return '"' + path.join(bowerPath, '..', '..', 'bin', 'bower') + '"';
+        });
+    });
   },
   _resolveModule: function(moduleName, options) {
     return resolve(moduleName, options);

--- a/lib/utils/config.js
+++ b/lib/utils/config.js
@@ -7,7 +7,9 @@ var RSVP = require('rsvp');
 var findByName = require('./find-by-name');
 var debug = require('debug')('ember-try:utils:config');
 
-function config(options) {
+var IMPLICIT_BOWER_VERSION = '^1.8.2';
+
+function getBaseConfig(options) {
   var relativePath = options.configPath || path.join('config', 'ember-try.js');
   var configFile = path.join(options.project.root, relativePath);
   var configData;
@@ -39,7 +41,38 @@ function config(options) {
   }
 }
 
+function config(options) {
+  return getBaseConfig(options).then(function(configData) {
+    return addImplicitBowerToScenarios(configData);
+  });
+}
+
 module.exports = config;
+
+function addImplicitBowerToScenarios(configData) {
+  configData.scenarios.forEach(function(scenario) {
+    if (!('bower' in scenario)) {
+      // Don't do anything for scenarios that don't include bower
+      return;
+    }
+
+    if ('npm' in scenario) {
+      var npm = scenario.npm;
+      if ((npm.dependencies && npm.dependencies.bower) ||
+          (npm.devDependencies && npm.devDependencies.bower)) {
+        // Dont' do anything for scenarios that already include bower in npm,
+        // either as a dependency or a dev dependency
+        return;
+      }
+    }
+
+    // add an implicit bower dev dependency to npm for this scenario
+    scenario.npm = scenario.npm || {};
+    scenario.npm.devDependencies = scenario.npm.devDependencies || {};
+    scenario.npm.devDependencies.bower = IMPLICIT_BOWER_VERSION;
+  });
+  return configData;
+}
 
 function mergeAutoConfigAndConfigFileData(autoConfig, configData) {
   configData = configData || {};
@@ -77,7 +110,12 @@ function defaultConfig() {
           dependencies: { } /* No dependencies needed as the
                                default is already specified in
                                the consuming app's bower.json */
-        }
+        },
+        npm: {
+          devDependencies: {
+            bower: IMPLICIT_BOWER_VERSION,
+          }
+        },
       },
       {
         name: 'ember-release',
@@ -85,7 +123,12 @@ function defaultConfig() {
           dependencies: {
             ember: 'release'
           }
-        }
+        },
+        npm: {
+          devDependencies: {
+            bower: IMPLICIT_BOWER_VERSION,
+          }
+        },
       },
       {
         name: 'ember-beta',
@@ -93,7 +136,12 @@ function defaultConfig() {
           dependencies: {
             ember: 'beta'
           }
-        }
+        },
+        npm: {
+          devDependencies: {
+            bower: IMPLICIT_BOWER_VERSION,
+          }
+        },
       },
       {
         name: 'ember-canary',
@@ -101,7 +149,12 @@ function defaultConfig() {
           dependencies: {
             ember: 'canary'
           }
-        }
+        },
+        npm: {
+          devDependencies: {
+            bower: IMPLICIT_BOWER_VERSION,
+          }
+        },
       }
     ]
   };
@@ -109,3 +162,4 @@ function defaultConfig() {
 
 // Used for internal testing purposes.
 module.exports._defaultConfig = defaultConfig;
+module.exports._addImplicitBowerToScenarios = addImplicitBowerToScenarios;

--- a/lib/utils/run.js
+++ b/lib/utils/run.js
@@ -6,7 +6,8 @@ var debug = require('debug')('ember-try:utils:run');
 
 function run(command, args, opts) {
   opts = opts || {};
-  opts.stdio = 'inherit';
+
+  opts.stdio = opts.stdio || 'inherit';
 
   if (process.env.SHUT_UP) {
     opts.stdio = 'ignore';

--- a/test/dependency-manager-adapters/bower-adapter-test.js
+++ b/test/dependency-manager-adapters/bower-adapter-test.js
@@ -121,33 +121,23 @@ describe('bowerAdapter', function() {
       return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._install();
     });
 
-    it('runs bower install with global bower if local bower is not found', function() {
-      var actualCommand, actualArgs, actualOpts;
-
+    it('rejects if local bower is not found', function() {
       var doNotFindLocalBower = function() {
+        return RSVP.reject('no local bower found');
+      };
+
+      var stubbedRun = function() {
         return RSVP.reject();
-      };
-
-      var findGlobalBower = function() {
-        return RSVP.resolve();
-      };
-
-      var stubbedRun = function(command, args, opts) {
-        actualCommand = command;
-        actualArgs = args;
-        actualOpts = opts;
-        return RSVP.resolve();
       };
 
       return new BowerAdapter({
         cwd: tmpdir,
         _findBowerPath: doNotFindLocalBower,
-        _checkForGlobalBower: findGlobalBower,
         run: stubbedRun
       })._install().then(function() {
-        expect(actualCommand).to.equal('bower');
-        expect(actualArgs).to.eql(['install', '--config.interactive=false']);
-        expect(actualOpts).to.have.property('cwd', tmpdir);
+        expect.fail(true, false, 'unreachable: _install promise rejects');
+      }, function(error) {
+        expect(error).to.equal('no local bower found');
       });
     });
 
@@ -280,34 +270,6 @@ describe('bowerAdapter', function() {
     it('returns the correct bower path', function() {
       return new BowerAdapter({ cwd: tmpdir })._findBowerPath().then(function(path) {
         expect(path).to.include('node_modules/bower/bin/bower');
-      });
-    });
-  });
-
-  describe('#_checkForGlobalBower()', function() {
-    it('runs bower -v to see if bower exists', function() {
-      let actualCommand, actualArgs, actualOpts;
-      var stubbedRun = function(command, args, opts) {
-        actualCommand = command;
-        actualArgs = args;
-        actualOpts = opts;
-        return RSVP.resolve();
-      };
-
-      return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._checkForGlobalBower().then(function() {
-        expect(actualCommand).to.equal('bower');
-        expect(actualArgs).to.eql(['-v']);
-        expect(actualOpts.stdio).to.equal('ignore');
-      });
-    });
-
-    it('throws if running bower -v fails', function() {
-      var stubbedRun = function() {
-        return RSVP.reject();
-      };
-
-      return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._checkForGlobalBower().catch(function(err) {
-        expect(err).to.match(/Bower must be installed either locally or globally to have bower scenarios/);
       });
     });
   });

--- a/test/dependency-manager-adapters/bower-adapter-test.js
+++ b/test/dependency-manager-adapters/bower-adapter-test.js
@@ -121,6 +121,36 @@ describe('bowerAdapter', function() {
       return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._install();
     });
 
+    it('runs bower install with global bower if local bower is not found', function() {
+      var actualCommand, actualArgs, actualOpts;
+
+      var doNotFindLocalBower = function() {
+        return RSVP.reject();
+      };
+
+      var findGlobalBower = function() {
+        return RSVP.resolve();
+      };
+
+      var stubbedRun = function(command, args, opts) {
+        actualCommand = command;
+        actualArgs = args;
+        actualOpts = opts;
+        return RSVP.resolve();
+      };
+
+      return new BowerAdapter({
+        cwd: tmpdir,
+        _findBowerPath: doNotFindLocalBower,
+        _checkForGlobalBower: findGlobalBower,
+        run: stubbedRun
+      })._install().then(function() {
+        expect(actualCommand).to.equal('bower');
+        expect(actualArgs).to.eql(['install', '--config.interactive=false']);
+        expect(actualOpts).to.have.property('cwd', tmpdir);
+      });
+    });
+
     it('runs bower install including managerOptions', function() {
       writeJSONFile('bower.json', fixtureBower);
       var stubbedRun = function(command, args) {
@@ -252,59 +282,32 @@ describe('bowerAdapter', function() {
         expect(path).to.include('node_modules/bower/bin/bower');
       });
     });
-
-    it('does not attempt to install bower if bower is found', function() {
-      var installCount = 0;
-      var stubbedResolveModule = function() {
-        return RSVP.resolve('blip/bloop/foo/lib/index.js');
-      };
-      var stubbedInstallBower = function() {
-        installCount++;
-      };
-      return new BowerAdapter({ cwd: tmpdir, _installBower: stubbedInstallBower, _resolveModule: stubbedResolveModule })._findBowerPath().then(function(path) {
-        expect(path).to.include('blip/bloop/foo/bin/bower');
-        expect(installCount).to.equal(0);
-      });
-    });
-
-    it('installs bower if bower is not found', function() {
-      var installCount = 0;
-      var resolveModuleCount = 0;
-      var stubbedResolveModule = function() {
-        resolveModuleCount++;
-        if (resolveModuleCount === 1) {
-          return RSVP.reject();
-        }
-        if (resolveModuleCount === 2) {
-          return RSVP.resolve('flip/flop/gloop/lib/index.js');
-        }
-      };
-
-      var stubbedInstallBower = function() {
-        installCount++;
-      };
-
-      return new BowerAdapter({ cwd: tmpdir, _installBower: stubbedInstallBower, _resolveModule: stubbedResolveModule })._findBowerPath().then(function(path) {
-        expect(path).to.include('flip/flop/gloop/bin/bower');
-        expect(installCount).to.equal(1);
-      });
-    });
   });
 
-  describe('#_installBower()', function() {
-    it('installs bower via npm', function() {
-      var command, args, opts;
-      var stubbedRun = function(c, a, o) {
-        command = c;
-        args = a;
-        opts = o;
+  describe('#_checkForGlobalBower()', function() {
+    it('runs bower -v to see if bower exists', function() {
+      let actualCommand, actualArgs, actualOpts;
+      var stubbedRun = function(command, args, opts) {
+        actualCommand = command;
+        actualArgs = args;
+        actualOpts = opts;
         return RSVP.resolve();
       };
-      return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._installBower().then(function() {
-        expect(command).to.equal('npm');
-        expect(args[0]).to.equal('install');
-        expect(args[1]).to.equal('bower@^1.3.12');
-        expect(opts).to.have.property('cwd', tmpdir);
+
+      return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._checkForGlobalBower().then(function() {
+        expect(actualCommand).to.equal('bower');
+        expect(actualArgs).to.eql(['-v']);
+        expect(actualOpts.stdio).to.equal('ignore');
+      });
+    });
+
+    it('throws if running bower -v fails', function() {
+      var stubbedRun = function() {
+        return RSVP.reject();
+      };
+
+      return new BowerAdapter({ cwd: tmpdir, run: stubbedRun })._checkForGlobalBower().catch(function(err) {
+        expect(err).to.match(/Bower must be installed either locally or globally to have bower scenarios/);
       });
     });
   });

--- a/test/tasks/try-each-test.js
+++ b/test/tasks/try-each-test.js
@@ -280,7 +280,6 @@ describe('tryEach', function() {
       });
     });
 
-
     it('fails scenarios when scenario\'s tests fail', function() {
       this.timeout(300000);
 

--- a/test/utils/config-test.js
+++ b/test/utils/config-test.js
@@ -9,6 +9,7 @@ var fixturePackage = require('../fixtures/package.json');
 var writeJSONFile = require('../helpers/write-json-file');
 var getConfig = require('../../lib/utils/config');
 var defaultConfig = getConfig._defaultConfig;
+var addImplicitBowerToScenarios = getConfig._addImplicitBowerToScenarios;
 
 var remove = RSVP.denodeify(fs.remove);
 var root = process.cwd();
@@ -36,6 +37,85 @@ describe('utils/config', function() {
     fs.writeFileSync('config/' + filename, contents, { encoding: 'utf8' });
   }
 
+  describe('addImplicitBowerToScenarios', function() {
+    it('adds an npm with a bower dev depencency for scenarios that have only bower', function() {
+      expect(addImplicitBowerToScenarios({
+        scenarios: [{
+          name: 'bower-only',
+          bower: { dependencies: {} },
+        }]
+      })).to.deep.equal({
+        scenarios: [{
+          name: 'bower-only',
+          bower: { dependencies: {} },
+          npm: { devDependencies: { bower: '^1.8.2' } },
+        }]
+      });
+    });
+
+    it('adds a bower dev dependency for scnearios that have bower and npm', function() {
+      expect(addImplicitBowerToScenarios({
+        scenarios: [{
+          name: 'bower-and-npm',
+          bower: { dependencies: {} },
+          npm: { devDependencies: { foo: 'latest' } },
+        }]
+      })).to.deep.equal({
+        scenarios: [{
+          name: 'bower-and-npm',
+          bower: { dependencies: {} },
+          npm: { devDependencies: { foo: 'latest', bower: '^1.8.2' } },
+        }]
+      });
+    });
+
+    it('does not add a bower dev dependency if a bower dev dependency is already present', function() {
+      expect(addImplicitBowerToScenarios({
+        scenarios: [{
+          name: 'bower-dev-dependency',
+          bower: { dependencies: {} },
+          npm: { devDependencies: { bower: '^1.8.2' } },
+        }]
+      })).to.deep.equal({
+        scenarios: [{
+          name: 'bower-dev-dependency',
+          bower: { dependencies: {} },
+          npm: { devDependencies: { bower: '^1.8.2' } },
+        }]
+      });
+    });
+
+    it('does not add a bower dev dependency if a bower dependency is already present', function() {
+      expect(addImplicitBowerToScenarios({
+        scenarios: [{
+          name: 'bower-dependency',
+          bower: { dependencies: {} },
+          npm: { dependencies: { bower: '^1.8.2' } },
+        }]
+      })).to.deep.equal({
+        scenarios: [{
+          name: 'bower-dependency',
+          bower: { dependencies: {} },
+          npm: { dependencies: { bower: '^1.8.2' } },
+        }]
+      });
+    });
+
+    it('does not add a bower dev dependency if bower is not present', function() {
+      expect(addImplicitBowerToScenarios({
+        scenarios: [{
+          name: 'no-bower',
+          npm: { devDependencies: { foo: 'latest' } },
+        }]
+      })).to.deep.equal({
+        scenarios: [{
+          name: 'no-bower',
+          npm: { devDependencies: { foo: 'latest' } },
+        }]
+      });
+    });
+  });
+
   it('uses specified options.configFile if present', function() {
     generateConfigFile('module.exports = { scenarios: [ { qux: "baz" }] };', 'non-default.js');
 
@@ -51,6 +131,29 @@ describe('utils/config', function() {
     return getConfig({ project: project }).then(function(config) {
       expect(config.scenarios).to.have.lengthOf(1);
       expect(config.scenarios[0].foo).to.equal('bar');
+    });
+  });
+
+  it('implicitly adds a bower dev dependency for npm for scenarios that include bower', function() {
+    generateConfigFile('module.exports = { scenarios: [ { bower: { dependencies: { foo: "latest" } } }] };');
+
+    return getConfig({ project: project }).then(function(config) {
+      expect(config.scenarios).to.have.lengthOf(1);
+      expect(config.scenarios[0]).to.deep.equal({
+        bower: { dependencies: { foo: 'latest' } },
+        npm: { devDependencies: { bower: '^1.8.2' } },
+      });
+    });
+  });
+
+  it('does not add a bower dependency for scenarios that do not include bower', function() {
+    generateConfigFile('module.exports = { scenarios: [ { npm: { dependencies: { foo: "latest" } } }] };');
+
+    return getConfig({ project: project }).then(function(config) {
+      expect(config.scenarios).to.have.lengthOf(1);
+      expect(config.scenarios[0]).to.deep.equal({
+        npm: { dependencies: { foo: 'latest' } },
+      });
     });
   });
 
@@ -98,23 +201,23 @@ describe('utils/config', function() {
       return getConfig({ project: project }).then(function(config) {
         expect(config).to.eql({
           scenarios: [
-            { name: 'default', bower: { dependencies: {} } },
+            { name: 'default', bower: { dependencies: {} }, npm: { devDependencies: { bower: '^1.8.2' } } },
             {
               name: 'ember-beta',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#beta' }, resolutions: { ember: 'beta' } },
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-canary',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#canary' }, resolutions: { ember: 'canary' } },
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-2.2.0',
               bower: { dependencies: { ember: '2.2.0' } },
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
           ],
         });
@@ -126,24 +229,23 @@ describe('utils/config', function() {
       return getConfig({ project: project, versionCompatibility: { ember: '1.13.0' } }).then(function(config) {
         expect(config).to.eql({
           scenarios: [
-            { name: 'default', bower: { dependencies: {} } },
+            { name: 'default', bower: { dependencies: {} }, npm: { devDependencies: { bower: '^1.8.2' } } },
             {
               name: 'ember-beta',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#beta' }, resolutions: { ember: 'beta' } },
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-canary',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#canary' }, resolutions: { ember: 'canary' } },
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-1.13.0',
               bower: { dependencies: { ember: '1.13.0' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             { foo: 'bar' },
           ],
@@ -155,26 +257,23 @@ describe('utils/config', function() {
       return getConfig({ project: project, versionCompatibility: { ember: '1.13.0' } }).then(function(config) {
         expect(config).to.eql({
           scenarios: [
-            { name: 'default', bower: { dependencies: {} } },
+            { name: 'default', bower: { dependencies: {} }, npm: { devDependencies: { bower: '^1.8.2' } } },
             {
               name: 'ember-beta',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#beta' }, resolutions: { ember: 'beta' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-canary',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#canary' }, resolutions: { ember: 'canary' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-1.13.0',
               bower: { dependencies: { ember: '1.13.0' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
           ],
         });
@@ -196,26 +295,27 @@ describe('utils/config', function() {
         expect(config).to.eql({
           bowerOptions: ['--allow-root=true'],
           scenarios: [
-            { name: 'default', bower: { dependencies: {} } },
+            {
+              name: 'default',
+              bower: { dependencies: {} },
+              npm: { devDependencies: { bower: '^1.8.2' } },
+            },
             {
               name: 'ember-beta',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#beta' }, resolutions: { ember: 'beta' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-canary',
               allowedToFail: true,
               bower: { dependencies: { ember: 'components/ember#canary' }, resolutions: { ember: 'canary' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
             {
               name: 'ember-2.2.0',
               bower: { dependencies: { ember: '2.2.0' } },
-
-              npm: { devDependencies: { 'ember-source': null } },
+              npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
             },
           ],
         });
@@ -230,26 +330,27 @@ describe('utils/config', function() {
         expect(config.useVersionCompatibility).to.equal(true);
         expect(config.bowerOptions).to.eql(['--allow-root=true']);
         expect(config.scenarios).to.eql([
-          { name: 'default', bower: { dependencies: {} } },
+          {
+            name: 'default',
+            bower: { dependencies: {} },
+            npm: { devDependencies: { bower: '^1.8.2' } },
+          },
           {
             name: 'ember-beta',
             allowedToFail: false,
             bower: { dependencies: { ember: 'components/ember#beta' }, resolutions: { ember: 'beta' } },
-
-            npm: { devDependencies: { 'ember-source': null } },
+            npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
           },
           {
             name: 'ember-canary',
             allowedToFail: true,
             bower: { dependencies: { ember: 'components/ember#canary' }, resolutions: { ember: 'canary' } },
-
-            npm: { devDependencies: { 'ember-source': null } },
+            npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
           },
           {
             name: 'ember-2.2.0',
             bower: { dependencies: { ember: '2.2.0' } },
-
-            npm: { devDependencies: { 'ember-source': null } },
+            npm: { devDependencies: { 'ember-source': null, bower: '^1.8.2' } },
           },
           { name: 'bar' },
         ]);

--- a/test/utils/dependency-manager-adapter-factory-test.js
+++ b/test/utils/dependency-manager-adapter-factory-test.js
@@ -17,7 +17,7 @@ describe('DependencyManagerAdapterFactory', function() {
       expect(adapters.length).to.equal(1);
     });
 
-    it('creates both adapters when it has both keys', function() {
+    it('creates both adapters, with npm first, when it has both keys', function() {
       var adapters = DependencyManagerAdapterFactory.generateFromConfig({ scenarios: [{ bower: {}, npm: {} }] }, 'here');
       expect(adapters[0].configKey).to.equal('npm');
       expect(adapters[1].configKey).to.equal('bower');

--- a/test/utils/scenario-manager-test.js
+++ b/test/utils/scenario-manager-test.js
@@ -32,7 +32,7 @@ describe('scenarioManager', function() {
   });
 
   describe('#changeTo', function() {
-    it('changes dependency sets on each of the managers, and concats results', function() {
+    it('changes dependency sets on each of the managers, in order, and concats results', function() {
       var fakeAdapters = [
         new CoreObject({
           changeToDependencySet: function() {

